discovery.md

```diff
@@ -101,18 +101,295 @@ be considered an event subscriber and a target user of this API.
+
+#### `/providers?matching=[search term]`
+
+Returns a list of all providers int he discovery system, optionally filtering on
+a provided search term (`matching`).
+
+##### matching
+
+* Type: `string`
+* Description: Search term that provides case insensitive match against
+  `provider` names. The parameter can match any portion of the provider name.
+* Constraints:
+  * OPTIONAL
+  * If present, must be non-empty
+* Examples:
+  * `"storage"`
+    * matches:
+      * `"Awesome Cloud Storage"`
+      * `"File storage system"`
+      * `"storage Storage STORAGE"`
+
+##### Returns
+
+A JSON array of `provider` entities.
+
+#### `/providers/{name}`
+
+Retrieves an individual provider entity by exact match on the `provider` name.
+
+##### Returns
+
+A `provider` entity as a JSON object.
+
+#### `/providers/{name}/types`
+
+Retrieves the details about the types that are offered by the specified
+`provider`.
+
+##### Returns
+
+A JSON array of `type` entities
+
+#### `/producer/{provider.name}/{type.name}`
+
+Retrieves the `producer` entity that specifies the information necessary to
+create subscriptions and to consume the events. The `provider.name` and
+`type.name` items in the request path correspond to the information that can
+be obtained via the `/provider` and `/type` API calls.
+
+##### Returns
+
+A `producer` entity as a JSON object.
+
+### Entity Specifications
+
+This section details the fields that make up each of the entities referenced
+earlier in this document.
+
+#### `provider` entity
+
+Used in discovery for enumerating the different providers represented in this
+discovery system.
+
+##### `provider` entity attributes
+
+###### provider
+
+- Type: `String`
+- Description: Identifies the provider of the event by nane. The producer string
+  SHOULD be human readable and can identify a top level producer system.
+- Constraints:
+  - REQUIRED
+  - MUST be a non-empty string
+- Examples:
+  - "GitHub.com"
+  - "Awesome Cloud Storage"
+  - "com.example.microservices.userlogin"
+
+###### description
+
+- Type: `String`
+- Description: Human readable description .
+- Constraints:
+  - OPTIONAL
+  - If present, MUST be a non-empty string
+
+###### provider_uri
+
+- Type: `URI`
+- Description: Absolute URI that provides a link back to the producer, or
+  documentation about the producer. This is for human consumption.
+- Constraints:
+  - OPTIONAL
+  - If present, MUST be a non-empty URI
+- Examples:
+  - "http://cloud.example.com/docs/blobstorage"
+
+
+##### `provider` entity examples
+
+```json
+{
+  "provider": "Cloud Storage Provider",
+  "description": "Blob storage in the cloud",
+  "provider_uri": "https://cloud.example.com/docs/storage"
+}
+```
+
+And a list of valid `provider` entities.
+
+```json
+[
+  {
+    "provider": "Cloud Storage Provider",
+    "description": "Blob storage in the cloud",
+    "provider_uri": "https://cloud.example.com/docs/storage"
+  },
+  {
+    "provider": "Cloud MySQL"
```

Comment:
I was going to suggest we use the provider_uri as the unique key, but since it's optional we may need to add an `id` to each one for that purpose.

Reply:
I was using `name` as the unique key. Any reason not to do that?